app/models/coupon.rb

```diff
@@ -1,4 +1,4 @@
 class Coupon < ApplicationRecord
   validates :amount, :name, presence: true
-  has_many :order_items, as: :source
+  has_many :order_items, as: :source, dependent: :destroy
```

Comment:
I don't think it should be on order item. If an order item is deleted you don't want to delete the product too. I think I said that product should have a deleted state.